pilot/pkg/networking/plugin/health/health.go

```diff
@@ -0,0 +1,134 @@
+	hcfilter "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/health_check/v2"
+	http_conn "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	xdsutil "github.com/envoyproxy/go-control-plane/pkg/util"
+	"github.com/gogo/protobuf/types"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
+)
+
+// Plugin implements Istio mTLS auth
+type Plugin struct{}
+
+// NewPlugin returns an instance of the health plugin
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// BuildHealthCheckFilter returns a HealthCheck filter.
+func BuildHealthCheckFilter(probe *model.Probe) *http_conn.HttpFilter {
+	return &http_conn.HttpFilter{
+		Name: xdsutil.HealthCheck,
+		Config: util.MessageToStruct(&hcfilter.HealthCheck{
+			PassThroughMode: &types.BoolValue{
+				Value: true,
+			},
+			Headers: []*envoy_api_v2_route.HeaderMatcher{
+				{
+					Name:  ":path",
+					Value: probe.Path,
+				},
+			},
+		}),
+	}
+}
+
+func buildHealthCheckFilters(filterChain *plugin.FilterChain, probes model.ProbeList) {
+	for _, probe := range probes {
+		filter := BuildHealthCheckFilter(probe)
+		if !containsHTTPFilter(filterChain.HTTP, filter) {
+			filterChain.HTTP = append(filterChain.HTTP, filter)
+		}
+	}
+}
+
+func containsHTTPFilter(array []*http_conn.HttpFilter, elem *http_conn.HttpFilter) bool {
+	for _, item := range array {
+		if reflect.DeepEqual(item, elem) {
+			return true
+		}
+	}
+	return false
+}
+
+// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
+// Can be used to add additional filters on the outbound path
+func (Plugin) OnOutboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	// TODO: implementation
+	return nil
+}
+
+// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+// Can be used to add additional filters (e.g., mixer filter) or add more stuff to the HTTP connection manager
+// on the inbound path
+func (Plugin) OnInboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	if in.Node.Type != model.Sidecar {
+		// Only care about sidecar.
+		return nil
+	}
+
+	if in.Node == nil {
+		return nil
+	}
+
+	if in.ServiceInstance == nil {
+		return nil
+	}
+
+	if mutable.Listener == nil {
+		return fmt.Errorf("listener not defined in mutable %v", mutable)
+	}
+
+	if len(mutable.Listener.FilterChains) != len(mutable.FilterChains) {
+		return fmt.Errorf("expected same number of filter chains in listener (%d) and mutable (%d)", len(mutable.Listener.FilterChains), len(mutable.FilterChains))
+	}
+
+	for i := range mutable.Listener.FilterChains {
+		if in.ListenerType == plugin.ListenerTypeHTTP {
+			buildHealthCheckFilters(&mutable.FilterChains[i], in.Env.WorkloadHealthCheckInfo(in.Node.IPAddress))
```

Comment:
You need to take the port into account here. IF the workloadHEalth check is defined on the same port as the inboundListener port and the inboundListener is a HTTP listener, then you setup the passthrough HTTP filter. Otherwise, you skip.

here is what is happening under the hood:
networking.core generates outbound listeners (http/tcp) for all service ports.
generates inbound listeners (http/tcp) for all endpoint ports/pod ports with endpoint.ip:endpoint.port
generates managementPort listeners for the podIP based on the health probes. These are TCP only listeners.
 if any of the managementPort listeners collide with the inboundListeners, the managementPort listener is skipped with a warning that health checks are supposed to be on distinct ports (not on same ports as service traffic). 

https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/listener.go#L130-L146

The reason we don't allow normal listener and health check listener on same port is because the mixer filter will intercept the inbound call and try to lookup the service associated with the pod. But pod->service association does not happen until the health checks pass. Since  mixer does not see any service associated with the pod, it "might" drop the request causing pod health check to fail (and thus the pod would never start/be part of service). 

Now, all said, the question is how will this even work with mixer enabled, given that if there is a port collision, pod health checks don't pass at all. 
I am starting to question those user reports that said their trace logs are filled with k8s health check requests. The only way it could have worked is when the user disabled mixer. In that case, what you have here makes sense.

Reply:
Ah ok - I think I understand. However I did have an example where the health check was on the same port, and it worked fine - but generated the health check traces. So possibly this is the only situation where it does occur.

I'm travelling for a couple of days but will try out the two scenarios and get back to you.